### ISSUE TITLE
Istio-ingressgateway getting stuck sometimes.

### ISSUE THREAD

**Issue by dguendisch** [gardener/gardener#4095]
**How to categorize this issue?**

/area networking
/area control-plane
/kind bug
/priority 2

**What happened**:
We noticed from time to time that some istio-ingressgateway pods stop working resulting in customers getting occasionally some `SSL_ERROR_SYSCALL` when e.g. curl'ing their shoot api servers. Simple deleting and thus restarting such istio pods resolved the issue.

What the failing istio pods have in common: the mem usage is somewhere >1.8GB and the envoy admin/master thread is constantly at 99% CPU load (check with `top -H` and then `Shift-v`)

```
top - 09:12:54 up 3 days,  2:30,  0 users,  load average: 6.02, 4.10, 3.84
Threads:  42 total,   2 running,  40 sleeping,   0 stopped,   0 zombie
%Cpu(s): 12.8 us,  3.5 sy,  0.0 ni, 81.9 id,  0.2 wa,  0.0 hi,  1.6 si,  0.0 st
KiB Mem : 64364876 total, 46168244 free, 15413088 used,  2783544 buff/cache
KiB Swap:        0 total,        0 free,        0 used. 50691548 avail Mem

    PID USER      PR  NI    VIRT    RES    SHR S %CPU %MEM     TIME+ COMMAND
     25 istio-p+  20   0  833256 455400   9136 R 99.9  0.7 478:23.29  `- envoy
     26 istio-p+  20   0  833256 455400   9136 S  0.0  0.7   0:00.00      `- default-executo
     27 istio-p+  20   0  833256 455400   9136 S  0.0  0.7   0:00.00      `- resolver-execut
     28 istio-p+  20   0  833256 455400   9136 S  0.0  0.7   0:03.06      `- grpc_global_tim
     29 istio-p+  20   0  833256 455400   9136 S  0.0  0.7   0:00.00      `- GrpcGoogClient
     30 istio-p+  20   0  833256 455400   9136 S  0.0  0.7   0:12.75      `- dog:main_thread
     31 istio-p+  20   0  833256 455400   9136 S  0.0  0.7   0:12.92      `- dog:workers_gua
     32 istio-p+  20   0  833256 455400   9136 S  4.0  0.7  91:57.52      `- wrk:worker_0
     33 istio-p+  20   0  833256 455400   9136 S  1.3  0.7  33:14.91      `- wrk:worker_1
     34 istio-p+  20   0  833256 455400   9136 S  0.0  0.7   0:00.00      `- GrpcGoogClient
     35 istio-p+  20   0  833256 455400   9136 S  1.3  0.7  63:31.89      `- wrk:worker_2
     36 istio-p+  20   0  833256 455400   9136 S  0.0  0.7   0:00.00      `- GrpcGoogClient
     37 istio-p+  20   0  833256 455400   9136 S  6.7  0.7 198:08.19      `- wrk:worker_3
     38 istio-p+  20   0  833256 455400   9136 S  0.0  0.7   0:00.00      `- GrpcGoogClient
     39 istio-p+  20   0  833256 455400   9136 S  0.0  0.7   0:00.00      `- GrpcGoogClient
     40 istio-p+  20   0  833256 455400   9136 S  0.0  0.7   0:00.20      `- AccessLogFlush
```

Furthermore we noticed that istio pods that are not "stuck yet" but already have some higher mem usage, are bogged exactly once per minute for some 10-20 secs (seems to depend on the mem usage), with the envoy admin thread at 99% load.
We furthermore confirmed that this is due to prometheus scraping at `/stats/prometheus` by manually hitting this endpoint from outside.

We conducted a final experiment and found that this admin thread seems to be needed for envoy to update its routing towards target pods:
Setup:
- we port-forward to one istio-ingressgateway pod and its 9443 port `k port-forward pod/istio-ingressgateway-776bd5b55f-pjshm 9443`
- we port-forward to this istos stats port `k port-forward pod/istio-ingressgateway-776bd5b55f-pjshm 15020`
- we constantly curl one shoot's kube-apiserver via this istio pod: `curl --resolve api.<shootname>.<project>.<some-domain>:9443:127.0.0.1 -kSs https://api.w1ij0v351f.miketest.shoot.canary.k8s-hana.ondemand.com:9443`
- we shoot down the kube-apiserver (to have a new pod started) and observe the pod and the respective `endpoint` resource and the curl command

Result:
- curl cannot reach the kapi once we shot it down
- after kube-apiserver was recreated, it takes at least some 10secs (maybe due to our 10-15secs liveness/readiness probe delay)
- when we did that action in "quiet" times (no `/stats/prometheus` queries reaching the istio pod):
	- then as soon as the `endpoint` was updated, the curl command heals and works again
- when we did the same action in "busy" times (us hitting the `/stats/prometheus` endpoint):
	- the `endpoint` update has no effect on the curl command, curl only heals and works again once the envoy admin thread is done with 99% CPU load from our `/stats/prometheus` invocation
- we tried that some 10 times with consistent results.

Conclusion:
The envoy admin thread is obviously needed to update envoys routing to the target pods (although we didn't expect this as we thought that the routing should happen on k8s/overlay network level only).
Prometheus puts the admin thread under great stress.
The envoy admin thread is a single thread and everything runs in a blocking way, hence while the stats endpoint is scraped, this admin thread doesn't do anything else.
See https://github.com/envoyproxy/envoy/issues/16425 for a very similar topic and more details on the admin thread.

**Environment**:

- Gardener version: 1.22 - 1.23

/cc @mvladev

**Comment by ialidzhikov** [comment]
Is this related to https://github.com/istio/istio/issues/29334 and https://github.com/gardener/gardener-extension-provider-alicloud/issues/181? Or is it another issue?

**Comment by dguendisch** [comment]
> Is this related to istio/istio#29334

Thanks for the pointer! I think this addresses in fact the same issue. And it seems there is already a fix on envoy side: https://github.com/istio/istio/issues/29334#issuecomment-825792362

**Comment by danielfoehrKn** [comment]
@mvladev updated istio from `1.8` to `1.9.5` with [this PR.](https://github.com/gardener/gardener/pull/4101/files).

Do I understand correctly that this version does not include yet any fixes regarding the envoy master process problems we have encountered?
This is just to make sure that we run an up-to-date version

EDIT: apparently the latest version should include [this commit](https://github.com/envoyproxy/envoy/pull/14954) which is potentially a fix (we have to observe that)

**Comment by danielfoehrKn** [comment]
In addition to fixing the problem with the main envoy process,  I think we should also adjust the resource requests of the istio-ingressgateway envoy pods.


The istio-ingressgateway is a critical component of a Seed cluster. It requires a lot of CPU time and memory depending on the API server traffic (probably mostly dependent on the number of Shoots on the Seed cluster).


- I think it makes sense to make the requested resources and limits of the istio-ingressgateway pods configurable per Seed cluster.
- We could also have a heuristic _defaulting_ the resource requests (cpu & memory) based on the number of Shoots if no explicit resources are set.
-  It should be a QoS garuanteed pod where the kubelet has the [CPU manager with static policy enabled](https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/#static-policy). 
   - Without CPU manager: only uses cgroupv1 CFS quotas `cpu.cfs\_quota\_us` (not get throtteled) and `cpu.shares` (guaranteed minimum CPU performance via CFS "fair shares" )
      - envoy processes will execute on various cores (comes with performance penalties) 
      - other pod container processes can still execute on the core where envoy threads are running
   - With CPU manager with static policy enabled (CPU pinning): the envoy process is the only pod container process that can execute on that CPU core (ensured via cgroupv1 `cpusets`). Other non-`kubepods` cgroup process can still execute on it - no exclusivity.

**Comment by danielfoehrKn** [comment]
@mvladev WDYT

**Comment by dguendisch** [comment]
fyi: I noticed on our canary landscape that `1.9.5` indeed either fixed or greatly reduced the problem. I don't see any istio pod getting high mem or high cpu after some days now. (will monitor the next days as well)

**Comment by danielfoehrKn** [comment]
@dguendisch  even on particular busy seed clusters the memory and cpu consumption is: 

```
k top pods
NAME                                   CPU(cores)   MEMORY(bytes)
istio-ingressgateway-98c468459-5p5hf   250m         223Mi
istio-ingressgateway-98c468459-92dzb   303m         371Mi
istio-ingressgateway-98c468459-gk46n   216m         126Mi
istio-ingressgateway-98c468459-lnv58   244m         152Mi
```
We currently define envoy as guaranteed QoS with 4Gi and 4 CPU (for all seeds). 
That is a heavy over-reservation 

>  we should make the requested resources and limits of the istio-ingressgateway pods configurable per Seed cluster.

If we really want to increase cpu performance of the envoy proxies, we should pin them via cpu.sets to certain cores using the CPU Manager feature.

@rfranzke can we add this issue to the quality focus?

**Comment by rfranzke** [comment]
Hm, given the many discussions and analyzations in the past weeks, I'd rather not add it right now and vote for first discussing the topic with @istvanballok @mvladev @dguendisch. If all agree, I'd think @mvladev would be most suitable for working on the topic.

**Comment by dguendisch** [comment]
Since the `1.9.5` update the consumption pattern changed indeed (however there are a few outliers depending on the seed: on a given landscape I noticed max usage cpu ~1200m and max mem usage 600Mi) . So yes, checking the current pattern and adapting to avoid waste makes sense in my opinion.

**Comment by dguendisch** [comment]
I think we can close this issue as we haven't observed istio getting stuck anymore after the update.
/close
